src/frequenz/sdk/microgrid/component_data.py

```diff
@@ -6,240 +6,243 @@
 License
 MIT
 """
+from __future__ import annotations
 
-from abc import ABC
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
 from datetime import datetime
-from typing import Tuple
+from typing import Optional, Tuple
 
 import frequenz.api.microgrid.microgrid_pb2 as microgrid_pb
 import pytz
 
 from .component_states import EVChargerCableState
 
 
-class _BaseComponentData(ABC):
-    """A private base class for strongly typed component data classes."""
+@dataclass(frozen=True)
+class ComponentData(ABC):
+    """A private base class for strongly typed component data classes.
 
-    _raw: microgrid_pb.ComponentData
+    Attributes:
+        component_id: the ID identifying this component in the microgrid
+        timestamp: the timestamp of when the data was measured.
+    """
 
-    @property
-    def component_id(self) -> int:
-        """Return the component id of the message.
+    component_id: int
+    timestamp: datetime
+    _raw: Optional[microgrid_pb.ComponentData] = field(default=None, init=False)
 
-        Returns:
-            Component id of the message.
-        """
-        return self._raw.id
-
-    @property
-    def timestamp(self) -> datetime:
-        """Return the timestamp of the message.
+    def _set_raw(self, raw: microgrid_pb.ComponentData) -> None:
+        """Store raw protobuf message.
 
-        Returns:
-            Timestamp of the message.
-        """
-        return self._raw.ts.ToDatetime(tzinfo=pytz.UTC)
-
-
-class MeterData(_BaseComponentData):
-    """A wrapper class for holding meter data."""
-
-    def __init__(self, raw: microgrid_pb.ComponentData) -> None:
-        """Create a `MeterData` instance.
+        Keep the dataclass frozen, but bypass it here to store the raw protobuf
+            message for in case it was needed later.
```

Comment:
Here also please be a bit more verbose, mention which alternatives we considered and why we decided for this approach.

Reply:
Done. Let me know if that's enough or sth else should be mentioned.

Reply:
LGTM, thanks!